net: add RAII socket and use it instead of bare SOCKET

vasild opened the pull request:
Introduce a class to manage the lifetime of a socket - when the object
that contains the socket goes out of scope, the underlying socket will
be closed.

In addition, the new `Sock` class has a `Send()`, `Recv()` and `Wait()`
methods that can be overridden by unit tests to mock the socket
operations.

The `Wait()` method also hides the
`#ifdef USE_POLL poll() #else select() #endif` technique from higher
level code.

vasild committed:
net: add RAII socket and use it instead of bare SOCKET

Introduce a class to manage the lifetime of a socket - when the object
that contains the socket goes out of scope, the underlying socket will
be closed.

In addition, the new `Sock` class has a `Send()`, `Recv()` and `Wait()`
methods that can be overridden by unit tests to mock the socket
operations.

The `Wait()` method also hides the
`#ifdef USE_POLL poll() #else select() #endif` technique from higher
level code.

vasild committed:
net: use Sock in InterruptibleRecv() and Socks5()

Use the `Sock` class instead of `SOCKET` for `InterruptibleRecv()` and
`Socks5()`.

This way the `Socks5()` function can be tested by giving it a mocked
instance of a socket.

Co-authored-by: practicalswift <practicalswift@users.noreply.github.com>

vasild commented:
The following two PRs would benefit from this:

* https://github.com/bitcoin/bitcoin/pull/19203 would only need to add the fuzzing test and the mocked/fuzzed `Sock` implementation (last commit from that PR).
* https://github.com/bitcoin/bitcoin/pull/20685 would become smaller. `Sock` originated there and that PR contains a copy of it. So if this PR gets merged, then the size of #20685 will be reduced.

MarcoFalke commented on src/sock.cpp:
```diff
@@ -0,0 +1,134 @@
+
+#include <compat.h>
+#include <netaddress.h>
+#include <netbase.h>
```
```suggestion
#include <netbase.h> // For CloseSocket
```

It seems this is the only Bitcoin Core related dependency. Are the other includes needed at all? If not, it would be good to move this module to `util`, which is the place for general utility stuff (not bitcoin related).

vasild replied:
No, the other includes are not needed (and I removed them). However there is a circular dependency now:

* netbase -> sock (`netbase.cpp` uses the `Sock` class, of course)
* sock -> netbase (`sock.cpp` uses `MillisToTimeval()` and `CloseSocket()` which are defined in `netbase.cpp`)

One way to resolve that is to put the `Sock` class in `netbase.{h,cpp}` and ditch `sock.{h,cpp}`.

Another one would be to move `MillisToTimeval()` and `CloseSocket()` outside of `netbase.cpp`, somewhere in `src/util/` and use them from both `netbase.cpp` and from `sock.cpp`.

Hmm...

vasild replied:
> One way to resolve that is to put the `Sock` class in `netbase.{h,cpp}`...

Did that.

vasild replied:
This was finally [changed](https://github.com/bitcoin/bitcoin/pull/20788#issuecomment-756754017) to:

> move MillisToTimeval() and CloseSocket() outside of netbase.cpp, somewhere in src/util/ and use them from both netbase.cpp and from sock.cpp

it looks much better now.

maflcko reviewed (commented):
left a style comment, no other review. Feel free to ignore.

jonatack reviewed (commented):
Concept ACK, was unable to build this pull @ 99b8e88a15cb12483f28079d0f91fe8cf361da56

<details><summary>build error output</summary><p>

```
$ gcc --version
gcc (Debian 10.2.1-1) 10.2.1 20201207

.../...

  AR       qt/libbitcoinqt.a
/usr/bin/ld: libbitcoin_common.a(libbitcoin_common_a-netbase.o): in function `std::_MakeUniq<Sock>::__single_object std::make_unique<Sock, unsigned int&>(unsigned int&)':
/usr/include/c++/10/bits/unique_ptr.h:962: undefined reference to `Sock::Sock(unsigned int)'
  CXXLD    qt/bitcoin-qt
  CXXLD    bitcoin-gui
/usr/bin/ld: libbitcoin_common.a(libbitcoin_common_a-netbase.o): in function `std::_MakeUniq<Sock>::__single_object std::make_unique<Sock, unsigned int&>(unsigned int&)':
/usr/include/c++/10/bits/unique_ptr.h:962: undefined reference to `Sock::Sock(unsigned int)'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:5503: bitcoin-node] Error 1
make[2]: *** Waiting for unfinished jobs....
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:5515: bitcoind] Error 1
/usr/bin/ld: libbitcoin_common.a(libbitcoin_common_a-netbase.o): in function `std::_MakeUniq<Sock>::__single_object std::make_unique<Sock, unsigned int&>(unsigned int&)':
/usr/include/c++/10/bits/unique_ptr.h:962: undefined reference to `Sock::Sock(unsigned int)'
/usr/bin/ld: libbitcoin_common.a(libbitcoin_common_a-netbase.o): in function `std::_MakeUniq<Sock>::__single_object std::make_unique<Sock, unsigned int&>(unsigned int&)':
/usr/include/c++/10/bits/unique_ptr.h:962: undefined reference to `Sock::Sock(unsigned int)'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:5521: qt/bitcoin-qt] Error 1
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:5499: bitcoin-gui] Error 1
make[2]: Leaving directory '/home/jon/projects/bitcoin/bitcoin/src'
make[1]: *** [Makefile:14882: all-recursive] Error 1
make[1]: Leaving directory '/home/jon/projects/bitcoin/bitcoin/src'
make: *** [Makefile:806: all-recursive] Error 1
```
</p></summary>

lontivero commented on src/netbase.cpp:
```diff
@@ -347,15 +349,15 @@ enum class IntrRecvError {
  *      SOCKET&, bool).
  */
-static IntrRecvError InterruptibleRecv(uint8_t* data, size_t len, int timeout, const SOCKET& hSocket)
+static IntrRecvError InterruptibleRecv(uint8_t* data, size_t len, int timeout, const Sock& sock)
```
Q: would it make sense to have this method as part of the `Sock`? (Something similar to `Sock::recv` but with a timeout parameter)

vasild replied:
Yes, indeed. That would be convenient together with a `send` method that retries upon temporary failures (see below).

I did not include such methods in this PR because it is self-sufficient as it is and achieves the initial purpose to introduce a mockable socket that can be used for fuzzing in https://github.com/bitcoin/bitcoin/pull/19203.

Possible future improvements (which I resisted adding in this PR):
* Add robust receiving method, like `InterruptibleRecv()` (or change that function to a method of the `Sock` class)
* Add robust sending method, like [`SendComplete()`](https://github.com/bitcoin/bitcoin/blob/2ae504c4b96013c87cc97909bdb00866fc4ab8a0/src/netbase.cpp#L1085) from https://github.com/bitcoin/bitcoin/pull/20685
* Use `Sock` in more places instead of the bare `SOCKET` - chase all places where `Sock::Release()` or `Sock::Get()` is called and see if the consumer can be changed to use `Sock` instead. E.g. in `CNode` constructor.

practicalswift commented:
Strong concept ACK

This abstraction would be extremely useful to have when fuzzing low-level networking code.

vasild commented:
99b8e88a1...408a9473b: moved the `Sock` class inside `netbase.cpp` to resolve a [circular dependency](https://github.com/bitcoin/bitcoin/pull/20788#discussion_r549799438).

Another option would be to keep the `Sock` class in its own file `src/sock.cpp` or `src/util/sock.cpp`. Since it uses `MillisToTimeval()` and `CloseSocket()` they would have to be moved out of `netbase.cpp`.

jonatack commented on src/netbase.h:
```diff
@@ -51,9 +54,130 @@ bool Lookup(const std::string& name, CService& addr, int portDefault, bool fAllo
+
+    /**
+     * recv(2) wrapper. Equivalent to `recv(this->Get(), buf, len, flags);`. Code that uses this
+     * wrapper can be unit-tested if this method is overriden by a mock Sock implementation.
```
nit here and L120, s/overriden/overridden/

vasild replied:
Fixed

lontivero commented on src/net.cpp:
```diff
@@ -437,51 +437,55 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
-    CAddress addr_bind = GetBindAddress(hSocket);
-    CNode* pnode = new CNode(id, nLocalServices, GetBestHeight(), hSocket, addrConnect, CalculateKeyedNetGroup(addrConnect), nonce, addr_bind, pszDest ? pszDest : "", conn_type);
+    CAddress addr_bind = GetBindAddress(sock->Get());
+    CNode* pnode = new CNode(id, nLocalServices, GetBestHeight(), sock->Release(), addrConnect,
+                             CalculateKeyedNetGroup(addrConnect), nonce, addr_bind,
+                             pszDest ? pszDest : "", conn_type);
```
nit: i would like to suggest rename `Sock::Release` as `Sock::GetAndRelease`. IMO it makes it just a tiny bit  more readable. Feel free to ignore this comment.
```c++
    CNode* pnode = new CNode(id, nLocalServices, GetBestHeight(), sock->GetAndRelease(), addrConnect,
                             CalculateKeyedNetGroup(addrConnect), nonce, addr_bind,
                             pszDest ? pszDest : "", conn_type);
```

vasild replied:
I agree `GetAndRelease()` is more descriptive, but I am trying to mimic `std::unique_ptr` and its [method which is called release()](https://en.cppreference.com/w/cpp/memory/unique_ptr/release).

jonatack reviewed (commented):
Tested approach ACK 408a9473bed6b3be82b124c171435247100b9d34 (modulo missing test coverage?)

- moving the Sock class to its own file and the common deps to a utility seems appealing, at least in principle, not sure in practice
- review of the diff would be easier and nicer, and there would be less unnecessary churn and risk of overlooking a regression, if unneeded formatting and line break changes were not added when only an argument in the function signature is being updated
- do you plan to add unit tests for the Sock class?

lontivero commented on src/netbase.cpp:
```diff
@@ -541,58 +530,170 @@ static bool Socks5(const std::string& strDest, int port, const ProxyCredentials
+    return recv(m_socket, static_cast<char*>(buf), len, flags);
+}
+
+bool Sock::Wait(const std::chrono::milliseconds& timeout, Event requested, Event* returned) const
```
The `returned` parameter is never used by any caller. It seems this is something that was added for future usage maybe. Does it make sense to have it at this point in time, I mean, if tomorrow it is needed then it could be added in a future PR.

vasild replied:
Removed, less code - the better.

lontivero reviewed (commented):
I've finished the review. Only two comments.

jnewbery commented:
Concept ACK. Is there any reason not to use the `Sock` object in `CNode`?

vasild commented:
408a9473b...8fc63e344: fixed a typo and removed unused parameter

vasild commented:
@jonatack 

> moving the Sock class to its own file and the common deps to a utility seems appealing

I will try this. Opinions, anybody? That would mean:
* Move `Sock` from `netbase.cpp` to `util/sock.cpp`
* Move `MillisToTimeval()` from `netbase.cpp` to `util/where?`
* Move `CloseSocket()` from `netbase.cpp` to `util/where?` (it uses `LogPrintf()`)

> do you plan to add unit tests for the Sock class?

I did not but any ideas are welcome. I think it may be required to create a TCP server and client and do some communication between them to check that `send()` is indeed being called by the `Send()` method. May be an overkill, given that the `Send()` method is 1 line.

I don't know how to reliably check if the socket was closed by the destructor, given that the same file descriptor number could be assigned to a newly created socket (or opened file) after we close it.

@jnewbery 

> Is there any reason not to use the `Sock` object in `CNode`?

No, other than minimizing the size of this PR. See also https://github.com/bitcoin/bitcoin/pull/20788#discussion_r549718454 :-)

lontivero commented:
I agree with moving it to its own file and with adding UTs for it. I am not sure a real socket is necessary for testing this abstraction, could it be enough to simply mock `send`, `recv` and `select`? Something it could be useful is to make sure the ownership of the `SOCKET` is always correct (`copy ctor`, `Release`, `operator =`)

btw, utACK 8fc63e344660f52a35bd97c1f881afdb1643b52a

vasild committed:
test: add Sock unit tests

vasild committed:
net: move CloseSocket() from netbase to util/sock

Move `CloseSocket()` (and `NetworkErrorString()` which it uses) from
`netbase.{h,cpp}` to newly added `src/util/sock.{h,cpp}`.

This is necessary in order to use `CloseSocket()` from a newly
introduced Sock class (which will live in `src/util/sock.{h,cpp}`).
`sock.{h,cpp}` cannot depend on netbase because netbase will depend
on it.

vasild committed:
net: move MillisToTimeval() from netbase to util/time

Move `MillisToTimeval()` from `netbase.{h,cpp}` to
`src/util/system.{h,cpp}`.

This is necessary in order to use `MillisToTimeval()` from a newly
introduced `src/util/sock.{h,cpp}` which cannot depend on netbase
because netbase will depend on it.

vasild commented:
8fc63e3... 3fa2a11:
* rebased due to conflicts
* moved `Sock` to its dedicated `src/util/sock.{cpp,h}`
* added unit tests

lontivero commented on src/test/sock_tests.cpp:
```diff
@@ -0,0 +1,150 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef WIN32
```
Q: Why are these tests disabled for Windows?

vasild replied:
They were because Windows does not have `socketpair()`, but since not all tests use this function I have now enabled the tests that don't.

lontivero commented on src/test/sock_tests.cpp:
```diff
@@ -0,0 +1,150 @@
+    BOOST_CHECK_EQUAL(sock->Release(), s);
+    delete sock;
+    BOOST_CHECK(!SocketIsClosed(s));
+    BOOST_REQUIRE_EQUAL(close(s), 0);
```
Q: Why not `CloseSocket` here?

vasild replied:
Right, changed to `CloseSocket()`.

lontivero commented on src/test/sock_tests.cpp:
```diff
@@ -0,0 +1,150 @@
+    Sock sock0(s[0]);
+    Sock sock1(s[1]);
+
+    std::thread waiter([&sock0]() { sock0.Wait(24h, Sock::RECV); });
```
Does `24h` means 24 hours? If that's the case it looks too much to me, even for a UT.

vasild replied:
Yes, 24 hours. I wanted to use `1y` (1 year), but `y` is not supported.

If this test detects a failure, it would result in a "test timeout in CI" which waits for 2 hours.

I don't know how else to check if `Wait()` works as expected. Notice - it does not signal to the caller whether a timeout occurred or the socket is ready for the event. It was possible before, but [we simplified](https://github.com/bitcoin/bitcoin/pull/20788#discussion_r550250637) it because nobody of the callers used that information. I think it is not worth restoring it for the sake of this test, what's your take?

If that functionality is provided then the test could do: "wait 1ms; ensure it caused timeout, send some data to the socket, wait 1ms; ensure it did not timeout, but signaled that the socket is ready".

lontivero replied:
> I think it is not worth restoring it for the sake of this test, what's your take?

I agree with you. I don't think it is a good idea to have code the sake of one test only.

lontivero commented on src/test/sock_tests.cpp:
```diff
@@ -0,0 +1,150 @@
+static SOCKET CreateSocket()
+{
+    const int s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    BOOST_REQUIRE(s != -1);
```
It could make sense to use `SOCKET_ERROR` here.

vasild replied:
Done, thanks!

lontivero reviewed (changes requested):
Mostly questions and one suggestion.

vasild commented:
3fa2a11f6...ed69709c8: hopefully fix CI and address some suggestions

lontivero commented:
utACK ed69709c8c86970f21b2a987a366737a01bc1c79

vasild commented:
ed69709c8...cc109a474: rebase due to conflicts

lontivero commented:
utACK cc109a47481263e1a7fb9b74f1aea57ea0129f8e

laanwj commented on src/Makefile.am:
```diff
@@ -238,6 +238,7 @@ BITCOIN_CORE_H = \
   util/rbf.h \
   util/ref.h \
   util/settings.h \
+  util/sock.h \
```
Please create new general utilities like this in `src/compat`, `src/support` or `src/util`, not in the source root.

vasild replied:
Hmm, the newly added `sock.h` is in `src/util`, so that's already the case, no?

laanwj replied:
Huh. Oops. I misread this. Seems good then.

laanwj commented:
Concept ACK

jonatack commented on src/netbase.cpp:
```diff
@@ -541,58 +523,52 @@ static bool Socks5(const std::string& strDest, int port, const ProxyCredentials
-        case SOCKS5Atyp::IPV4: recvr = InterruptibleRecv(pchRet3, 4, SOCKS5_RECV_TIMEOUT, hSocket); break;
-        case SOCKS5Atyp::IPV6: recvr = InterruptibleRecv(pchRet3, 16, SOCKS5_RECV_TIMEOUT, hSocket); break;
+        case SOCKS5Atyp::IPV4: recvr = InterruptibleRecv(pchRet3, 4, SOCKS5_RECV_TIMEOUT, sock); break;
+        case SOCKS5Atyp::IPV6: recvr = InterruptibleRecv(pchRet3, 16, SOCKS5_RECV_TIMEOUT, sock); break;
```
The diff could be smaller, faster, easier and safer to review without needless renamings mixed into the diff. Maybe open a separate pull for the renamings and the formatting changes.

vasild replied:
I split this commit in two: variable type change from `SOCKET` to `Sock` + variable rename from `hSocket` to `sock`.

It is good to have the implied convention of naming `SOCKET` variables `hSocket` (old code) or `socket` (new code) and `Sock` variables `sock`.

jonatack commented on src/net.cpp:
```diff
@@ -428,51 +428,55 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
+    CAddress addr_bind = GetBindAddress(sock->Get());
+    CNode* pnode = new CNode(id, nLocalServices, sock->Release(), addrConnect,
+                             CalculateKeyedNetGroup(addrConnect), nonce, addr_bind,
+                             pszDest ? pszDest : "", conn_type);
```
The diff could be smaller, faster, easier and safer to review without needless formatting churn in the diff. There are quite a few functions with formatting churn. Maybe open a separate pull for the renamings and the formatting changes to make review safer and easier.

vasild replied:
Joined this back to one long line.

jonatack reviewed (commented):
Verified debug build is clean and tests pass at each commit. I'm not very confident in reviewing here with the formatting changes and various renamings mixed into the signifcant code changes. This could be a smaller, easier and safer diff to review.

vasild committed:
style: rename hSocket to sock

In the arguments of `InterruptibleRecv()`, `Socks5()` and
`ConnectThroughProxy()` the variable `hSocket` was previously of type
`SOCKET`, but has been changed to `Sock`. Thus rename it to `sock` to
imply its type, to distinguish from other `SOCKET` variables and to
abide to the coding style wrt variables' names.

vasild commented on src/util/system.cpp:
```diff
@@ -1351,3 +1351,54 @@ std::pair<int, char**> WinCmdLineArgs::get()
+    if (ret) {
+        LogPrintf("Socket close failed: %d. Error: %s\n", socket, NetworkErrorString(WSAGetLastError()));
+    }
+    socket = INVALID_SOCKET;
```
Changed the commit to be move-only and left the variable name as `hSocket`.

vasild commented on src/util/time.h:
```diff
@@ -57,4 +60,14 @@ std::string FormatISO8601DateTime(int64_t nTime);
+/**
+ * Convert milliseconds to a struct timeval for e.g. select.
+ */
+struct timeval MillisToTimeval(const std::chrono::milliseconds& ms);
```
Changed.

vasild commented on src/util/sock.h:
```diff
@@ -0,0 +1,111 @@
+
+#include <compat.h> // SOCKET
+
+#include <chrono> // std::chrono::milliseconds
```
Those comments are useful, if they are present, during refactoring when moving code around. Unfortunately they are not widely used, so I do not insist on their presence. Removed.

vasild commented:
cc109a474...bb5231615: stylistic changes to address suggestions by @jonatack, thanks!

jonatack commented:
Thanks @vasild. I didn't see your latest notifications as I have been adjusting to the new `ghwatch.py` script, will re-review.

laanwj commented:
It seems more logical to me to move the `CloseSocket` and `NetworkErrorString` functions to `sock.h`, not `system.h`? Also do we still need an explicit `CloseSocket` instead of `Sock::Reset`?

Cursory code review ACK otherwise.

vasild commented:
bb5231615...7b6e4ee30: move `CloseSocket()` and `NetworkErrorString()` to `util/sock.{h,cpp}` instead of to `util/system.{h,cpp}`.

@laanwj,

> It seems more logical to me to move the `CloseSocket` and `NetworkErrorString` functions to `sock.h`, not `system.h`?

Done.

> ~Also do we still need an explicit `CloseSocket` instead of `Sock::Reset`?~

Actually this question makes sense - we can remove `CloseSocket()` altogether once its callers have switched to using the newly introduced `Sock` from this PR. There are about 10 places where `CloseSocket()` is called - should be possible (not in this PR).

laanwj commented:
@vasild Thanks! Yes, agree with not getting rid of `CloseSocket` in this PR, but glad to hear there's a path to removing the loose function.

Code review ACK e8e372a43122366fab7d6c7a45e05d4ad18d6da7

jonatack commented on src/test/sock_tests.cpp:
```diff
@@ -0,0 +1,149 @@
+{
+    // Notice that if another thread is running and creates its own socket after `s` has been
+    // closed, it may get assigned the same file descriptor number. In this case our test will
+    // wronly pretend that the socket is not closed.
```
```suggestion
    // wrongly pretend that the socket is not closed.
```
or "incorrectly"

vasild replied:
Yeah, that was wron. Fixed.

jonatack commented on src/test/sock_tests.cpp:
```diff
@@ -0,0 +1,149 @@
+static bool SocketIsClosed(const SOCKET& s)
+{
+    // Notice that if another thread is running and creates its own socket after `s` has been
+    // closed, it may get assigned the same file descriptor number. In this case our test will
```
```suggestion
    // closed, it may be assigned the same file descriptor number. In this case, our test will
```

vasild replied:
Fixed.

jonatack reviewed (commented):
ACK e8e372a43122366fab7d6c7a45e05d4ad18d6da7 rebased to master, debug build + code review at each commit, at head ran the new sock tests and the other unit tests locally and ran bitcoind while observing peer behavior--operation appears to be nominal

Thanks for the recent updates that I found made review much easier. Below a couple of doc nits in the unit test that can be ignored or done in the follow-up if you're inclined.

laanwj commented:
I think this is ready for merge, but would be nice to fix @jonatack's comment typos.

vasild commented:
e8e372a43...615ba0eb9: fix typos in a comment

jonatack commented:
Thanks @vasild, back to my computer in a half hour to re-ACK.

laanwj commented:
Re-ACK 615ba0eb96cf131364c1ceca9d3dedf006fa1e1c
Checked that only change is the comment.

jonatack reviewed (approved):
re-ACK 615ba0eb96cf131364c1ceca9d3dedf006fa1e1c

jonatack commented:
(Review and comment made from Work with GitHub pull requests

USAGE
  gh pr <command> [flags]

CORE COMMANDS
  checkout:   Check out a pull request in git
  checks:     Show CI status for a single pull request
  close:      Close a pull request
  comment:    Create a new pr comment
  create:     Create a pull request
  diff:       View changes in a pull request
  list:       List and filter pull requests in this repository
  merge:      Merge a pull request
  ready:      Mark a pull request as ready for review
  reopen:     Reopen a pull request
  review:     Add a review to a pull request
  status:     Show status of relevant pull requests
  view:       View a pull request

FLAGS
  -R, --repo [HOST/]OWNER/REPO   Select another repository using the [HOST/]OWNER/REPO format

INHERITED FLAGS
  --help   Show help for command

ARGUMENTS
  A pull request can be supplied as argument in any of the following formats:
  - by number, e.g. "123";
  - by URL, e.g. "https://github.com/OWNER/REPO/pull/123"; or
  - by the name of its head branch, e.g. "patch-1" or "OWNER:patch-1".

EXAMPLES
  $ gh pr checkout 353
  $ gh pr create --fill
  $ gh pr view --web

LEARN MORE
  Use 'gh <command> <subcommand> --help' for more information about a command.
  Read the manual at https://cli.github.com/manual as an experiment.)

practicalswift commented:
@vasild Thanks for making our socket handling mockable. This will be extremely useful when fuzzing!

Similar to what you've done for TCP connections it would be really useful to be able to globally mock out DNS lookup to be able to have the responses be generated by a `FuzzedDataProvider` driven class instead of the ordinary lookup facility provided by the OS.

If you want to address this mocking need too you can find some inspiration in #19415 :)

vasild commented:
Just for reference: previous incarnations of this PR had the `Sock::Wait()` method report to the caller whether one of the requested events happened or a timeout occurred. This was [removed](https://github.com/bitcoin/bitcoin/pull/20788#discussion_r550250637) because none of the callers needed it.

However this is needed in #20685, so that PR contains a tiny [commit](https://github.com/bitcoin/bitcoin/pull/20685/commits/8bc35b9cb1b38386c7234f251959fdb2f121ac07) to re-introduce that functionality.
